File: flink_jobs/job1_gps_normalizer.py
import json
import os
from pyflink.datastream import StreamExecutionEnvironment, MapFunction
from pyflink.common.serialization import SimpleStringSchema
from pyflink.datastream.connectors.kafka import KafkaSource, KafkaOffsetsInitializer, KafkaSink, KafkaRecordSerializationSchema
from pyflink.datastream.connectors.cassandra import CassandraSink
from pyflink.common import WatermarkStrategy

# --- 1. Define the Zone Lookup Logic ---
class CasablancaZoneMapper(MapFunction):
    def __init__(self):
        # We define your CSV data directly here for high performance in Task 01
        self.zones = [
            {"id": 1, "name": "Sidi Belyout", "lon": [-7.625, -7.595], "lat": [33.590, 33.620]},
            {"id": 2, "name": "Maarif", "lon": [-7.660, -7.615], "lat": [33.565, 33.595]},
            {"id": 3, "name": "Anfa", "lon": [-7.710, -7.660], "lat": [33.585, 33.630]},
            {"id": 4, "name": "Hay Hassani", "lon": [-7.730, -7.670], "lat": [33.530, 33.585]},
            {"id": 5, "name": "Mers Sultan", "lon": [-7.625, -7.595], "lat": [33.555, 33.590]},
            {"id": 6, "name": "Ain Chock", "lon": [-7.660, -7.610], "lat": [33.510, 33.565]},
            {"id": 7, "name": "Hay Mohammadi", "lon": [-7.595, -7.560], "lat": [33.585, 33.615]},
            {"id": 8, "name": "Sidi Bernoussi", "lon": [-7.540, -7.480], "lat": [33.600, 33.645]},
            {"id": 9, "name": "Ain Sebaa", "lon": [-7.560, -7.510], "lat": [33.595, 33.630]},
            {"id": 10, "name": "Roches Noires", "lon": [-7.595, -7.560], "lat": [33.605, 33.630]},
            {"id": 11, "name": "Sidi Moumen", "lon": [-7.540, -7.480], "lat": [33.560, 33.605]},
            {"id": 12, "name": "El Fida", "lon": [-7.615, -7.585], "lat": [33.565, 33.585]},
            {"id": 13, "name": "Mechouar", "lon": [-7.615, -7.600], "lat": [33.575, 33.590]},
            {"id": 14, "name": "Ben Msik", "lon": [-7.585, -7.550], "lat": [33.555, 33.580]},
            {"id": 15, "name": "Sbata", "lon": [-7.585, -7.540], "lat": [33.535, 33.565]},
            {"id": 16, "name": "Moulay Rachid", "lon": [-7.550, -7.500], "lat": [33.540, 33.580]}
        ]

    def map(self, value):
        event = json.loads(value)
        lon, lat = event['lon'], event['lat']
        speed = event.get('speed', 0)

        # 1. Validation: Speed Check
        if speed > 150: return None

        # 2. Validation: Zone Lookup & Anonymization
        for zone in self.zones:
            if (zone['lon'][0] <= lon <= zone['lon'][1]) and \
               (zone['lat'][0] <= lat <= zone['lat'][1]):
                
                # ENFORCE ANONYMIZATION: Replace raw with Centroid
                event['zone_id'] = zone['id']
                event['city'] = "Casablanca"
                event['lon'] = sum(zone['lon']) / 2
                event['lat'] = sum(zone['lat']) / 2
                return event
        
        return None # Filter out if not in any zone

def run_gps_normalizer():
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(1)

    # 1. Kafka Source
    source = KafkaSource.builder() \
        .set_bootstrap_servers("localhost:9092") \
        .set_topics("raw.gps") \
        .set_group_id("flink-job1-gps") \
        .set_starting_offsets(KafkaOffsetsInitializer.earliest()) \
        .set_value_only_deserializer(SimpleStringSchema()) \
        .build()

    stream = env.from_source(source, WatermarkStrategy.no_watermarks(), "Kafka Source")

    # 2. Apply Transformations
    normalized_stream = stream.map(CasablancaZoneMapper()).filter(lambda x: x is not None)

    # 3. Sink to Kafka (processed.gps)
    kafka_sink = KafkaSink.builder() \
        .set_bootstrap_servers("localhost:9092") \
        .set_record_serializer(
            KafkaRecordSerializationSchema.builder()
                .set_topic("processed.gps")
                .set_value_serialization_schema(SimpleStringSchema())
                .build()
        ).build()
    
    normalized_stream.map(lambda x: json.dumps(x)).sink_to(kafka_sink)

    # 4. Sink to Cassandra (taasim.vehicle_positions)
    # We map the dictionary to a tuple that matches the Cassandra Table schema:
    # PRIMARY KEY ((city, zone_id), event_time)
    cassandra_stream = normalized_stream.map(lambda x: (
        x['city'], x['zone_id'], x['timestamp'], x['taxi_id'], 
        x['lat'], x['lon'], float(x['speed']), x['status']
    ))

    CassandraSink.add_sink(cassandra_stream) \
        .set_host("127.0.0.1") \
        .set_query("INSERT INTO taasim.vehicle_positions (city, zone_id, event_time, taxi_id, lat, lon, speed, status) VALUES (?, ?, ?, ?, ?, ?, ?, ?)") \
        .build()

    env.execute("TaaSim Job 1: GPS Normalizer")

if __name__ == '__main__':
    run_gps_normalizer()

File: flink_jobs/src/main/java/com/taasim/flink/job1/Job1GpsNormalizer.java
package com.taasim.flink.job1;

import com.taasim.flink.job1.functions.ParseGpsEventFn;
import com.taasim.flink.job1.functions.ValidationAndLateFilterFn;
import com.taasim.flink.job1.functions.ZoneMappingBroadcastFn;
import com.taasim.flink.job1.model.GpsNormalizedEvent;
import com.taasim.flink.job1.model.GpsRawEvent;
import com.taasim.flink.job1.model.ZoneDefinition;
import com.taasim.flink.job1.util.ZoneMappingLoader;
import java.nio.charset.StandardCharsets;
import java.sql.Timestamp;
import java.time.Duration;
import java.util.List;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.SerializationSchema;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.api.common.state.MapStateDescriptor;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple8;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.connector.base.DeliveryGuarantee;
import org.apache.flink.connector.kafka.sink.KafkaRecordSerializationSchema;
import org.apache.flink.connector.kafka.sink.KafkaSink;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend;
import org.apache.flink.streaming.connectors.cassandra.CassandraSink;
import org.apache.flink.streaming.api.CheckpointingMode;
import org.apache.flink.streaming.api.datastream.BroadcastStream;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class Job1GpsNormalizer {
    private static final String DEFAULT_KAFKA_BOOTSTRAP = "kafka:29092";
    private static final String DEFAULT_SOURCE_TOPIC = "raw.gps";
    private static final String DEFAULT_SINK_TOPIC = "processed.gps";
    private static final String DEFAULT_CASSANDRA_HOST = "cassandra";
    private static final int DEFAULT_CASSANDRA_PORT = 9042;
    private static final String DEFAULT_CITY = "casablanca";
    private static final String DEFAULT_CHECKPOINT_DIR =
            "s3a://taasim/raw/kafka-archive/flink-checkpoints/job1/";

    public static void main(String[] args) throws Exception {
        final ParameterTool params = ParameterTool.fromArgs(args);

        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        buildJob(env, params);

        env.execute("job1-gps-normalizer");
    }

        static void buildJob(StreamExecutionEnvironment env, ParameterTool params) throws Exception {
        if (env == null) {
            throw new IllegalArgumentException("env must not be null");
        }
        if (params == null) {
            throw new IllegalArgumentException("params must not be null");
        }

        final String kafkaBootstrap = params.get("kafka-bootstrap-servers", DEFAULT_KAFKA_BOOTSTRAP);
        final String sourceTopic = params.get("source-topic", DEFAULT_SOURCE_TOPIC);
        final String sinkTopic = params.get("sink-topic", DEFAULT_SINK_TOPIC);
        final String cassandraHost = params.get("cassandra-host", DEFAULT_CASSANDRA_HOST);
        final int cassandraPort = params.getInt("cassandra-port", DEFAULT_CASSANDRA_PORT);
        final String city = params.get("city", DEFAULT_CITY);
        final String checkpointDir = params.get("checkpoint-dir", DEFAULT_CHECKPOINT_DIR);
        final long checkpointIntervalMs = params.getLong("checkpoint-interval-ms", 60_000L);

        env.getConfig().setGlobalJobParameters(params);
        env.getConfig().setAutoWatermarkInterval(1_000L);

        env.enableCheckpointing(checkpointIntervalMs);
        env.getCheckpointConfig().setCheckpointingMode(CheckpointingMode.AT_LEAST_ONCE);
        env.getCheckpointConfig().setCheckpointStorage(checkpointDir);
        env.getCheckpointConfig().setMinPauseBetweenCheckpoints(30_000L);
        env.getCheckpointConfig().setMaxConcurrentCheckpoints(1);
        env.setStateBackend(new EmbeddedRocksDBStateBackend());

        final KafkaSource<String> rawGpsSource =
                KafkaSource.<String>builder()
                        .setBootstrapServers(kafkaBootstrap)
                        .setTopics(sourceTopic)
                        .setGroupId("flink-job1-gps")
                        .setStartingOffsets(OffsetsInitializer.earliest())
                        .setValueOnlyDeserializer(new SimpleStringSchema())
                        .build();

        final DataStream<String> rawJson =
                env.fromSource(rawGpsSource, WatermarkStrategy.noWatermarks(), "Kafka raw.gps");

        final DataStream<GpsRawEvent> parsed = rawJson.flatMap(new ParseGpsEventFn()).name("parse-json");

        final WatermarkStrategy<GpsRawEvent> watermarkStrategy =
                WatermarkStrategy.<GpsRawEvent>forBoundedOutOfOrderness(Duration.ofMinutes(3))
                        .withTimestampAssigner((event, ts) -> event.eventTimeMillis);

        final DataStream<GpsRawEvent> watermarked =
                parsed.assignTimestampsAndWatermarks(watermarkStrategy).name("watermarks");

        final SingleOutputStreamOperator<GpsRawEvent> validated =
                watermarked.process(new ValidationAndLateFilterFn()).name("validate-and-late-filter");

        final List<ZoneDefinition> zones = ZoneMappingLoader.loadZonesFromClasspath();
        final DataStream<ZoneDefinition> zoneStream =
                zones.isEmpty()
                        ? env.fromElements(new ZoneDefinition()).filter(z -> false).name("zone-mapping-empty")
                        : env.fromCollection(zones).name("zone-mapping");

        final MapStateDescriptor<Integer, ZoneDefinition> zoneStateDesc =
                new MapStateDescriptor<>(
                        "zones",
                        Types.INT,
                        TypeInformation.of(new TypeHint<ZoneDefinition>() {}));

        final BroadcastStream<ZoneDefinition> zoneBroadcast = zoneStream.broadcast(zoneStateDesc);

        final DataStream<GpsNormalizedEvent> normalized =
                validated
                        .connect(zoneBroadcast)
                        .process(new ZoneMappingBroadcastFn(zoneStateDesc, city))
                        .name("zone-map-and-anonymize");

        final DataStream<Tuple8<String, Integer, Timestamp, String, Double, Double, Float, String>> cassandraRows =
                normalized
                        .map(
                                e ->
                                        Tuple8.of(
                                                e.city,
                                                e.zoneId,
                                                new Timestamp(e.eventTimeMillis),
                                                e.taxiId,
                                                e.lat,
                                                e.lon,
                                                e.speedKmh,
                                                e.status))
                        .returns(
                                Types.TUPLE(
                                        Types.STRING,
                                        Types.INT,
                                        Types.SQL_TIMESTAMP,
                                        Types.STRING,
                                        Types.DOUBLE,
                                        Types.DOUBLE,
                                        Types.FLOAT,
                                        Types.STRING))
                        .name("to-cassandra-rows");

        final String insertCql =
                "INSERT INTO taasim.vehicle_positions (city, zone_id, event_time, taxi_id, lat, lon, speed, status) "
                        + "VALUES (?, ?, ?, ?, ?, ?, ?, ?);";

        CassandraSink.addSink(cassandraRows).setQuery(insertCql).setHost(cassandraHost, cassandraPort).build();

        final SerializationSchema<Tuple2<String, String>> kafkaKeySerializer =
                element -> element.f0.getBytes(StandardCharsets.UTF_8);
        final SerializationSchema<Tuple2<String, String>> kafkaValueSerializer =
                element -> element.f1.getBytes(StandardCharsets.UTF_8);

        final KafkaSink<Tuple2<String, String>> processedGpsSink =
                KafkaSink.<Tuple2<String, String>>builder()
                        .setBootstrapServers(kafkaBootstrap)
                        .setDeliverGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
                        .setRecordSerializer(
                                KafkaRecordSerializationSchema.<Tuple2<String, String>>builder()
                                        .setTopic(sinkTopic)
                                        .setKeySerializationSchema(kafkaKeySerializer)
                                        .setValueSerializationSchema(kafkaValueSerializer)
                                        .build())
                        .build();

        normalized
                .map(e -> Tuple2.of(e.taxiId, e.toJson()))
                .returns(Types.TUPLE(Types.STRING, Types.STRING))
                .name("to-processed-gps-json")
                .sinkTo(processedGpsSink)
                .name("kafka-processed.gps");
    }
}
